Stochastic Backscatter Model for Grey Area Mitigation in Hybrid RANS-LES Simulations

AngPass opened the pull request:
## Proposed Changes
Hybrid RANS-LES methods aim to leverage the strengths of both Reynolds-Averaged Navier-Stokes (RANS) approaches and Large Eddy Simulation (LES), applying RANS near solid walls for computational efficiency and LES in outer flow regions for improved fidelity. A persistent challenge in these techniques is the appearance of the so-called grey area—the transitional region between RANS and LES zones—which is characterized by the absence of resolved turbulence. This leads to degraded accuracy in capturing free shear layers. The Stochastic Backscatter Model (Kok, 2017) aims to mitigate this issue. The model introduces stochastic forcing terms into the momentum and total energy equations that emulate the backscatter of energy from unresolved (subgrid) scales to resolved scales. The random forcing is formulated as the curl of a stochastic vector potential, say $\underline\xi$. The latter must be constructed using a three-step procedure:

1. A vector of three mutually-uncorrelated normally-distributed random variables is generated at each node.
2. A Laplacian smoothing in space (Passariello, 2025) is performed on each component of this vector.
3. The smoothed fields are used as source terms in three stochastic differential equations (also known as Langevin equations). The latter are solved for the components of the vector potential $\underline\xi$.
4. The curl of $\underline\xi$ is added to the momentum and total energy equations as source term. This can be done modifying the definition of the modeled stress tensor.

**References:** 
- Kok, J. C. (2017). A stochastic backscatter model for grey-area mitigation in detached eddy simulations. Flow, Turbulence and Combustion, 99(1), 119-150.
- Passariello, A. (2025). Stochastic Backscatter Model for Unstructured Solvers. arXiv preprint arXiv:2508.17958.

### Roadmap
1. Modifying the definition of the modeled stress tensor using a stochastic contribution. As preliminary step, the vector of uncorrelated variables (point 1 in the procedure above) is employed.
2. Performing a smoothing on the uncorrelated variables. This requires the solution of a Laplace equation at the beginning of each step. 
3. Implementing three additional equations for the components of the stochastic vector potential. These equations require the use of symmetry-preserving central scheme for the discretization of the convective terms (Kok, 2017).

## Related Work
None.

## PR Checklist
- [X] I am submitting my contribution to the develop branch.
- [X] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [X] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [x] I used the pre-commit hook to prevent dirty commits and used `pre-commit run --all` to format old commits.
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [x] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

AngPass committed:
Preliminary implementation: uncorrelated stochastic variables

The Reynolds stress tensor definition is modified to include a random contribution, which is expressed as the curl of a normally-distributed stochastic vector potential.

AngPass committed:
Merge branch 'su2code:develop' into develop

AngPass committed:
Add solution of Langevin equations

Add Langevin equations to Spalart-Allmaras solver (uncorrelated random source term) + add stochastic contribution to Reynolds stress tensor (using conservative variables from Langevin equations).

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Fix sanitizer warnings and regression failures

Fix sanitizer warnings and regression test failures

- Address issues reported by clang sanitizers.
- Adjust implementation to ensure regression tests pass.

AngPass committed:
Minor fixes

- Enhance readability.
- Ensure consistent declaration types.

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Ensure consistency with DES

- Extend implementation to 2D configurations.
- Improve robustness.

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Boundary conditions for Langevin equations

Enforce boundary conditions in Langevin equations.

AngPass committed:
Use symmetry-preserving scheme for Langevin equations

- Implement a skew-symmetric scheme for the discretization of the convective terms in the Langevin equations.
- Fix the seed for the generation of the random source terms at the beginning of the simulation.

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Add option for simulating DIHT

- Add the flag for the simulation of the Decaying Isotropic Homogeneous Turbulence.
- Generate the initial velocity field matching the experimental data by Comte-Bellot & Corrsin.
- Include a preliminary test case.

AngPass committed:
Fix options for DIHT

- Implement flag to enforce LES in the whole domain.
- Add option to set the LES filter width to a user-specified value.
- Add velocity divergence to volume outputs.
- Remove internal generation of initial velocity field for simulating the Decaying Isotropic Homogeneous Turbulence (DIHT).

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Consistent estimation of turb. kinetic energy (for SA)

- Add consistent evaluation of the turbulent kinetic energy in the random source term appearing in the momentum equations.
- Add backscatter intensity coefficient in the configuration file.
- Add random initialization of the Langevin variables.

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Add Laplacian smoothing (Langevin equations)

- Add Laplacian smoothing of source terms in Langevin equations.
Remark: pseudo-time integration is employed (residuals are printed on screen with fixed frequency, the maximum number of time iterations can be defined by the user).

AngPass committed:
Fix redundancy in CTurbSAVariable.hpp

- Fix redundancy in virtual member definition.

AngPass committed:
Add stochastic source to turbulence model equation

- Add stochastic source term to Spalart-Allmaras turbulence model equation (to ensure exchange of modeled and resolved kinetic energy).

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
SOR algorithm for Laplacian smoothing

- Replace dual-time integration with Successive Over-Relaxation for Laplacian smoothing.
- Initialize stochastic vector potential as equal to the stochastic source terms in Langevin equations.
- Add random source term to main balance equations in LES zones exclusively.
- Blend RANS and LES turbulence timescales using the LES sensor.

AngPass committed:
Merge branch 'su2code:develop' into develop

AngPass committed:
Correct scaling of stochastic source terms in Langevin eqs.

- Scale source terms in Langevin equations using Bessel functions to preserve the variance.
-  Compute Bessel integral at the beginning of the simulation for optimization.
- Add variance monitoring to screen output.

AngPass committed:
Merge branch 'su2code:develop' into develop

AngPass committed:
Merge branch 'su2code:develop' into develop

AngPass committed:
Add LD2 scheme for the incompressible solver

- Add LD2 discretization for the inviscid terms in the main balance equations.
Remarks:
-- Valid only for the incompressible flow solver, with constant density fluid model.
-- JST scheme must be selected in the config file. The novel option LD2_SCHEME must be set to YES.

AngPass committed:
Merge branch 'develop' of https://github.com/AngPass/SU2 into develop

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Fix LD2 for periodic boundaries

- Retrieve classic second-order central scheme at periodic boundaries.

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Include stochastic source term in turb. equation

Add option to include stochastic contribution to turbulence model equation.

Commit:
Minor fixes

- Use logarithmic approximation for Bessel function.
- Compute relative variance of the stochastic field for verification.

Commit:
Add time-averaged skin friction coefficient

- Add option to print time-averaged skin friction coefficient in volume output.

AngPass committed:
Minor fix

- Minor syntax fix in CConfig.cpp

AngPass committed:
Minor fix

- Minor syntax fix in CConfig.cpp

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Merge branch 'develop' into develop

Commit:
Add fields to VOLUME_OUTPUT

- Include time-averaged skin friction coefficient and instantaneous energy backscatter ratio into volume output fields.
- Fix boundary conditions for the implicit smoothing of the stochastic source term in Langevin equations.
- Fix computation of the source term in turbulence model equation.
- Fix computation of the subgrid kinetic energy (divide eddy viscosity by flow density).
- Diagonalize Jacobian in turbulence equations for numerical robustness.

Commit:
Fix LD2 option in config file

- Allow LD2 option only for JST scheme.

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Merge branch 'develop' into develop

Commit:
Enhance numerical robustness

- Add option to include diagnostics of the stochastic source term
  smoothing.
- Add relaxation factor for the random forcing term in the main balanca
  equations.
- Add modeled and stochastic Reynolds stresses to volume output files.

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Merge branch 'develop' into develop

Commit:
Enhance numerical efficiency

- Random generators removed from static variables.
- Random generators re-initialized at every time step using
  deterministic seeds.

AngPass committed:
Merge branch 'develop' into develop

Commit:
Fix bug in viscous flux computation

- Account for zero DES length scale in ghost cells.
- Fix threshold (=0.9) for the activation of the backscatter model.

pcarruscag reviewed (commented):
Some comments to start, I'll look more into the TurbSolver changes later

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Merge branch 'develop' into develop

Commit:
First major revision

- Add LD2 scheme to CONV_NUM_METHOD enum.
- Change scaling of the backscatter term.
- Increase the size of residual and Jacobian in turb_sources.hpp.
- Remove dynamic allocation in CTurbSASolver.cpp.
- Add helper function to compute relaxed backscatter intensity
  coefficient.
- Include pressure extrapolation in LD2 scheme formulation.
- Make code style homogeneous.

AngPass commented on Common/src/CConfig.cpp:
```diff
@@ -2958,9 +2958,42 @@ void CConfig::SetConfig_Options() {
+  addDoubleOption("LES_FILTER_WIDTH", LES_FilterWidth, -1.0);
+
+  /* DESCRIPTION: Specify if the LD2 scheme must be employed (incompressible flows, combined with JST discretization). */
+  addBoolOption("LD2_OPTION", LD2_Scheme, false);
```
Done.

AngPass commented on SU2_CFD/include/numerics/CNumerics.hpp:
```diff
@@ -639,6 +648,34 @@ class CNumerics {
+   */
+  template<class Mat, class Scalar, class Vector>
+  NEVERINLINE static void ComputeStochReynStress(size_t nDim, Scalar density, Scalar tke,
+                                                 Vector rndVec, Mat& stochReynStress, Scalar Cmag) {
```
Done.

AngPass commented on SU2_CFD/include/numerics/CNumerics.hpp:
```diff
@@ -837,6 +874,27 @@ class CNumerics {
+  inline void SetStochVar(su2double val_stochvar_i, su2double val_stochvar_j, unsigned short iDim) {
+    stochVar_i[iDim] = val_stochvar_i;
+    stochVar_j[iDim] = val_stochvar_j;
+  }
```
Done.

AngPass commented on SU2_CFD/include/numerics/scalar/scalar_convection.hpp:
```diff
@@ -52,6 +52,7 @@ class CUpwScalar : public CNumerics {
   const FlowIndices idx;            /*!< \brief Object to manage the access to the flow primitives. */
   su2double a0 = 0.0;               /*!< \brief The maximum of the face-normal velocity and 0. */
   su2double a1 = 0.0;               /*!< \brief The minimum of the face-normal velocity and 0. */
+  su2double qij = 0.0;              /*!< \brief The face-normal velocity (Langevin equations). */
```
Corrected.

AngPass commented on SU2_CFD/include/numerics/scalar/scalar_convection.hpp:
```diff
@@ -140,6 +141,13 @@ class CUpwScalar : public CNumerics {
+      for (unsigned short iDim = 0; iDim < nDim; iDim++) {
+        qij += 0.5 * (V_i[idx.Density()]*V_i[iDim + idx.Velocity()] + V_j[idx.Density()]*V_j[iDim + idx.Velocity()]) * Normal[iDim];
+      }
+    }
```
Done.

AngPass commented on SU2_CFD/include/numerics/turbulent/turb_convection.hpp:
```diff
@@ -59,6 +64,18 @@ class CUpwSca_TurbSA final : public CUpwScalar<FlowIndices> {
+    }
     Flux[0] = a0*ScalarVar_i[0] + a1*ScalarVar_j[0];
     Jacobian_i[0][0] = a0;
     Jacobian_j[0][0] = a1;
```
Only the eddy-viscosity-related component is overwritten. I implemented the code this way to perform proper inizialization of the Jacobian matrix.

AngPass commented on SU2_CFD/include/numerics/turbulent/turb_sources.hpp:
```diff
@@ -74,6 +74,8 @@ class CSourceBase_TurbSA : public CNumerics {
   su2double Residual, *Jacobian_i;
   su2double Jacobian_Buffer; /*!< \brief Static storage for the Jacobian (which needs to be pointer for return type). */
+  su2double* ResidSB = nullptr;
+  su2double** JacobianSB_i = nullptr; 
```
Done.

AngPass commented on SU2_CFD/include/numerics/turbulent/turb_sources.hpp:
```diff
@@ -112,23 +114,128 @@ class CSourceBase_TurbSA : public CNumerics {
+    ResidSB[0] = Residual;
+
+    for (unsigned short iVar = 0; iVar < nVar; iVar++ ) {
+      for (unsigned short jVar = 0; jVar < nVar; jVar++ ) {
```
When the backscatter model is activated, three additional stochastic differential equations are solved in CTurbSASolver, so `nVar=4`.

AngPass commented on SU2_CFD/include/solvers/CFVMFlowSolverBase.inl:
```diff
@@ -467,6 +468,30 @@ void CFVMFlowSolverBase<V, R>::Viscous_Residual_impl(unsigned long iEdge, CGeome
+    if (max(lesMode_i, lesMode_j) > threshold) {
+      tke_i = pow(eddy_visc_i/DES_length_i, 2);
+      tke_j = tke_i;
+      if (geometry->nodes->GetDomain(jPoint)) tke_j = pow(eddy_visc_j/DES_length_j, 2);
```
I modified this part. Now `tke` is estimated using the rate-of-strain magnitude. No parallelization problem should occur.

AngPass commented on SU2_CFD/include/solvers/CTurbSASolver.hpp:
```diff
@@ -39,7 +39,13 @@
+  su2double* Res_Wall = nullptr;
+  su2double** Jacobian_i = nullptr;
+  su2double* nu_tilde_inturb = nullptr;
+  su2double* nu_tilde_WF = nullptr;
```
Done.

AngPass commented on SU2_CFD/include/variables/CTurbSAVariable.hpp:
```diff
@@ -40,7 +41,11 @@ class CTurbSAVariable final : public CTurbVariable {
+  MatrixType stochSource;
+  MatrixType stochSource_old;
   VectorType Vortex_Tilting;
+  VectorType besselIntegral;
```
Done. Sorry.

AngPass commented on SU2_CFD/include/variables/CVariable.hpp:
```diff
@@ -397,6 +398,61 @@ class CVariable {
+   * \param[in] iDim - Dimension index.
+   * \param[in] val_stochSource_old - Old value of source term in Langevin equations.
+   */
+  inline virtual void SetLangevinSourceTermsOld(unsigned long iPoint, unsigned short iDim, su2double val_stochSource_old) {}
```
I moved `GetLangevinSourceTermsOld` and `SetLangevinSourceTermsOld` to CTurbVariable.hpp since they are called inside `CTurbSASolver.cpp` only.

AngPass commented on SU2_CFD/src/numerics/flow/convection/centered.cpp:
```diff
@@ -326,6 +328,31 @@ CNumerics::ResidualType<> CCentJSTInc_Flow::ComputeResidual(const CConfig* confi
+                                       velGrad_i[iDim][2] * d_ij[2]);
+      Velocity_j[iDim] -= alpha_LD2 * (velGrad_j[iDim][0] * d_ij[0] +
+                                       velGrad_j[iDim][1] * d_ij[1] +
+                                       velGrad_j[iDim][2] * d_ij[2]);
```
Indeed, I missed the pressure extrapolation. Sorry. 
The artificial dissipation is the same as used for the JST scheme, even though the fourth-order coefficient should be reduced. See the reference below.

_Carlsson, M., Davidson, L., Peng, S. H., & Arvidson, S. (2023). Investigation of low-dissipation low-dispersion schemes for incompressible and compressible flows in scale-resolving simulations. Computers & Fluids, 251, 105741._

AngPass commented on SU2_CFD/src/numerics/flow/flow_diffusion.cpp:
```diff
@@ -445,10 +456,32 @@ CNumerics::ResidualType<> CAvgGrad_Flow::ComputeResidual(const CConfig* config)
+      intensityCoeff = SBS_Cmag * (1.0 - exp(- currentTime / (timeScale*SBS_RelaxFactor)));
+    }
+    ComputeStochReynStress(nDim, Mean_PrimVar[nDim+2], Mean_turb_ke,
+                           Mean_StochVar, stochReynStress, intensityCoeff);
```
Done.

AngPass commented on SU2_CFD/src/output/CFlowCompOutput.cpp:
```diff
@@ -234,6 +234,8 @@ void CFlowCompOutput::SetVolumeOutputFields(CConfig *config){
   }
 
+  AddVolumeOutput("VELOCITY_DIVERGENCE", "Velocity_Divergence", "DERIVED", "Divergence of the velocity field");
+
```
I removed this volume output. I forgot it was still present in the code. I used it temporarily to assess the noise produced by the backscatter term.

AngPass commented on SU2_CFD/src/output/CFlowCompOutput.cpp:
```diff
@@ -325,6 +327,13 @@ void CFlowCompOutput::LoadVolumeData(CConfig *config, CGeometry *geometry, CSolv
+  for (unsigned short iDim = 0; iDim < nDim; iDim++) {
+    divVel += VelocityGradient[iDim][iDim];
+  }
+  SetVolumeOutputValue("VELOCITY_DIVERGENCE", iPoint, divVel);
```
See above conversation.

AngPass commented:
@pcarruscag I am currently encountering a bug that I have not been able to resolve on my own.
The issue seems to originate from the handling of MPI communications of periodic limiters. A segmentation fault arises when the backscatter model is active, periodic boundary conditions are present and the flag `MUSCL_TURB` is set to `YES`. I have not identified the root cause yet.
Any suggestions or guidance would be greatly appreciated.

**UPDATE:** it seems that allocating `Sol_Min` and `Sol_Max` in `CSolver::InitiatePeriodicComms` as `auto *Sol_Min   = new su2double[nVar]; auto *Sol_Max   = new su2double[nVar]` instead of `auto *Sol_Min   = new su2double[nPrimVarGrad]; auto *Sol_Max   = new su2double[nPrimVarGrad]` solves the issue.

AngPass committed:
Merge branch 'develop' into develop

Commit:
Minor fixes

- Redefine default values for backscatter parameters.
- Introduce RANS-LES blending factor for source terms in Langevin
  equations.
- Fix bug in MPI communication of slope limiters between periodic faces.

AngPass committed:
Merge branch 'develop' into develop

Commit:
Add new features

- Add the option to apply the backscatter model only in a bounded box.
- Add the option to enforce backscatter where shielding function exceeds
  user-defined value.
- Correct bug in LD2 implementation.
- Make style consistent and update configuration template.

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Merge branch 'develop' into develop

Commit:
Redefine scaling of stochastic source term

- Modify the scaling of the stochastic forcing to include the modeled
  viscosity.
- Fix the computation of the stochastic tensor.
- Add the MPI communication of the DES lengthscale.

Commit:
Update config template

AngPass committed:
Delete TestCases/backscatter/DIHT directory

Remove outdated configuration file.

Commit:
Include stochastic variables in restart file

- Add the stochastic variables (from Langevin equations) to the restart
  file.
- Optimize the definition of the points lying inside the box where the
  backscatter model is active.
- Limit the turbulence time scale in RANS regions (Langevin equations).
- Add test case: spatially-developing mixing layer.

AngPass committed:
Merge branch 'develop' into develop

AngPass commented:
@pcarruscag I added the cfg file of a test case (i.e., the spatially-developing mixing layer). Should I open a separate pull request under _Test Cases_ to upload the grid, or is it fine to include everything here?

pcarruscag commented:
Large files in testcases please.

AngPass committed:
Merge branch 'develop' into develop

Commit:
Change proposed test case

AngPass commented:
> Large files in testcases please.

@pcarruscag I uploaded the grid for the proposed test case (i.e., the backward-facing step). I had to compress it. The reference pull request in the TestCases repo is [https://github.com/su2code/TestCases/pull/187](https://github.com/su2code/TestCases/pull/187).

The code is now ready for review.

pcarruscag commented:
Add a regression test to parallel_regression.py, you'll probably need to do something custom to decompress the file.
It would be good to make it a bit smaller if possible too.

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -0,0 +1,153 @@
+  } else {
+    double t = 3.75 / abx;
+    double poly = 0.39894228 + t * (0.01328592 + t * (0.00225319 + t * (-0.00157565 + t * (0.00916281 + t * (-0.02057706 + t * (0.02635537 + t * (-0.01647633 + t * 0.00392377)))))));
+    return abx - 0.5 * log(abx) + log(poly);
```
```suggestion
    return abx - log(sqrt(abx) * poly);
```

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -0,0 +1,153 @@
+ * \return Standard normal random number (mean=0, stddev=1).
+ */
+inline double HashToNormal(uint64_t x) {
+    constexpr double pi = 3.14159265358979323846;
```
already defined in our constants

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -0,0 +1,153 @@
+    return splitmix64(x);
+}
+
+/*!
+ * \brief Convert a 64-bit hash into a uniform double in (0,1].
+ * Uses the top 53 bits of the hash to fill the mantissa of a double.
+ * Ensures the result is never zero, suitable for Box-Muller transform.
+ * \param[in] x 64-bit hash.
+ * \return Uniform double in the interval (0,1].
+ */
+inline double HashToUniform(uint64_t x) {
+    constexpr double inv53 = 1.0 / 9007199254740992.0; // 1/2^53
+    uint64_t uInt = x >> 11;                           // top 53 bits
+    return (uInt + 1) * inv53;                         // map to (0,1]
+}
```
Can you add some unit tests to ensure that this doesn't go out of bounds due to round-off?

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -0,0 +1,153 @@
+ * \param[in] x Argument of Bessel funtion.
+ * \return Value of Bessel function.
+ */
+inline double GetBesselZero(double x) {
```
No reason why these can't be AD-friendly right?
```suggestion
inline su2double GetBesselZero(su2double x) {
```

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -0,0 +1,153 @@
+ * \param[in] beta_z Argument in z-direction.
+ * \return Value of the integral.
+ */
+inline double GetBesselIntegral(double beta_x, double beta_y, double beta_z) {
```
same for su2double

pcarruscag commented on Common/include/CConfig.hpp:
```diff
@@ -1099,6 +1099,19 @@ class CConfig {
+  su2double stochFdThreshold;          /*!< \brief Shielding function lower threshold for application of Stochastic Backscatter Model. */
+  su2double stochSourceRelax;          /*!< \brief Relaxation factor for stochastic source term generation (Stochastic Backscatter Model). */
+  bool enforceLES;                     /*!< \brief Option to enforce LES mode in hybrid RANS-LES simulations. */
+  su2double LES_FilterWidth;           /*!< \brief LES filter width for hybrid RANS-LES simulations. */
```
Make some structs for this please

pcarruscag commented on Common/include/CConfig.hpp:
```diff
@@ -9604,6 +9617,42 @@ class CConfig {
+   * \brief Get the LES Filter Width.
+   * \return Value of LES Filter Width.
+   */
+  su2double GetLES_FilterWidth(void) const { return LES_FilterWidth; }
```
Then there is only one function to return a reference to the struct instead of many functions.
This is what we've been doing for other turbulence models.

pcarruscag commented on SU2_CFD/include/numerics/turbulent/turb_convection.hpp:
```diff
@@ -59,6 +63,17 @@ class CUpwSca_TurbSA final : public CUpwScalar<FlowIndices> {
+          Jacobian_j[iVar][jVar] = (iVar == jVar) ? 0.5*(a0+a1) : 0.0;
+        }
+      }
+    }
```
If you initialize the Jacobian buffers in the constructor with zeros, here you just set the diagonal in the same loop of the flux

pcarruscag commented on SU2_CFD/include/numerics/turbulent/turb_convection.hpp:
```diff
@@ -59,6 +63,17 @@ class CUpwSca_TurbSA final : public CUpwScalar<FlowIndices> {
    * \param[in] config - Definition of the particular problem.
    */
   void FinishResidualCalc(const CConfig* config) override {
+    if (config->GetStochastic_Backscatter() && config->GetSBS_Ctau() > 0.0) {
```
If Ctau is 0, are there no convective fluxes?
Seems very discontinuous.

AngPass replied:
The idea is the following: the three stochastic differential equations (SDEs) solved in the turbulence solver are needed only to generate a time‑correlated random source term.
If we set $C_\tau=0$, meaning no temporal correlation, then those additional SDEs are no longer necessary.
In that limit the model simply uses the instantaneous quantity coming from the smoothing operator, without integrating any extra differential equations.
So the behaviour is not a discontinuity in the physics, but rather the natural reduction of the model when the correlation time goes to zero.

pcarruscag commented on SU2_CFD/include/numerics/turbulent/turb_sources.hpp:
```diff
@@ -119,17 +191,19 @@ class CSourceBase_TurbSA : public CNumerics {
     /*--- Setup the Jacobian pointer, we need to return su2double** but we know
      * the Jacobian is 1x1 so we use this trick to avoid heap allocation. ---*/
-    Jacobian_i = &Jacobian_Buffer;
+    //Jacobian_i = &Jacobian_Buffer;
```
```suggestion
```

pcarruscag commented on SU2_CFD/include/numerics/turbulent/turb_sources.hpp:
```diff
@@ -168,11 +248,13 @@ class CSourceBase_TurbSA : public CNumerics {
       const su2double Ji_3 = Ji_2 * var.Ji;
 
       var.fv1 = Ji_3 / (Ji_3 + var.cv1_3);
+      if (config->GetStochastic_Backscatter() && lesMode_i > 0.0) var.fv1 = 1.0;
```
Do this within the framework of the different turbulence variants please.
What flavours of SA can it be combined with?

AngPass replied:
According to Deck’s  formulation $^1$, the near‑wall SA functions must be disabled in LES mode to accelerate turbulence development following the RANS-to-LES switch. The modifications are valid across the entire SA family.

$^1$ Deck, Sébastien et al. “Zonal Detached Eddy Simulation of a spatially developing flat plate turbulent boundary layer.” _Computers & Fluids_ 48 (2011): 1-15.

pcarruscag commented on SU2_CFD/include/numerics/turbulent/turb_sources.hpp:
```diff
@@ -168,11 +248,13 @@ class CSourceBase_TurbSA : public CNumerics {
        * From NASA turb modeling resource and 2003 paper. ---*/
       var.fv2 = 1 - ScalarVar_i[0] / (nu + ScalarVar_i[0] * var.fv1);
+      if (config->GetStochastic_Backscatter() && lesMode_i > 0.0) var.fv2 = 0.0;
       var.d_fv2 = -(1 / nu - Ji_2 * var.d_fv1) / pow(1 + var.Ji * var.fv1, 2);
```
The derivatives are not consistent with your changes

pcarruscag commented on SU2_CFD/include/numerics/CNumerics.hpp:
```diff
@@ -640,6 +652,62 @@ class CNumerics {
+   * \param[in] Cmag - Stochastic backscatter intensity coefficient.
+   * \param[out] stochReynStress - Stochastic tensor (to be added to the Reynolds stress tensor).
+   */
+  NEVERINLINE static void ComputeStochReynStress(su2double density, su2double tke, su2double rndVec[3], 
```
Why not inline a small function?

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -1650,6 +1691,85 @@ void CFlowOutput::LoadVolumeDataScalar(const CConfig* config, const CSolver* con
+      const auto vorticity = Node_Flow->GetVorticity(iPoint);
+      const su2double energy_backscatter = R_xy*vorticity[2] - R_xz*vorticity[1] + R_yz*vorticity[0];
+      const su2double energy_backscatter_ratio = energy_backscatter / (energy_res_to_mod + 1e-10);
+      SetVolumeOutputValue("ENERGY_BACKSCATTER_RATIO", iPoint, energy_backscatter_ratio);
```
Make some helper functions for all this.
Isn't the strain magnitude already computed and stored somewhere?
If not make a helper function to share the code for it.

pcarruscag commented on SU2_CFD/src/numerics/flow/flow_diffusion.cpp:
```diff
@@ -944,10 +988,26 @@ CNumerics::ResidualType<> CGeneralAvgGrad_Flow::ComputeResidual(const CConfig* c
+    su2double intensityCoeff = ComputeStochRelaxFactor(config);
+    ComputeStochReynStress(Mean_PrimVar[nDim+2], Mean_turb_ke_estim,
+                           Mean_StochVar, intensityCoeff, stochReynStress);
+  }
```
No duplication please, make helper functions for pieces of code that repeat.

pcarruscag commented on SU2_CFD/src/numerics/flow/convection/centered.cpp:
```diff
@@ -344,6 +349,32 @@ CNumerics::ResidualType<> CCentJSTInc_Flow::ComputeResidual(const CConfig* confi
+    }
+    MeanPressure += 0.5 * alpha_LD2 * ((pressGrad_i[0] - pressGrad_j[0]) * d_ij[0] +
+                                       (pressGrad_i[1] - pressGrad_j[1]) * d_ij[1] +
+                                       (pressGrad_i[2] - pressGrad_j[2]) * d_ij[2]);
```
use the geometry toolbox

pcarruscag commented on SU2_CFD/src/numerics/flow/convection/centered.cpp:
```diff
@@ -344,6 +349,32 @@ CNumerics::ResidualType<> CCentJSTInc_Flow::ComputeResidual(const CConfig* confi
+      for (iDim = 0; iDim < nDim; iDim++) {
+        velGrad_i[iDim][jDim] = PrimVar_Grad_i[iDim+1][jDim];
+        velGrad_j[iDim][jDim] = PrimVar_Grad_j[iDim+1][jDim];
+      }
```
Why are you copying into local variables that are used right away?

pcarruscag commented on SU2_CFD/src/numerics/flow/convection/centered.cpp:
```diff
@@ -326,6 +328,9 @@ CNumerics::ResidualType<> CCentJSTInc_Flow::ComputeResidual(const CConfig* confi
     Velocity_j[iDim]    = V_j[iDim+1];
+  }  
+
+  for (iDim = 0; iDim < nDim; iDim++) {
```
```suggestion
```

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -54,6 +55,18 @@ CTurbSASolver::CTurbSASolver(CGeometry *geometry, CConfig *config, unsigned shor
+    } else {
+      SU2_MPI::Error("Stochastic Backscatter Model available for 3D flows only.", CURRENT_FUNCTION);
+    }
+  }
```
Do this in CConfig post processing instead of here so that the config can be checked in dry-run mode

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -503,19 +556,21 @@ void CTurbSASolver::BC_HeatFlux_Wall(CGeometry *geometry, CSolver **solver_conta
 }
 
+
+
```
```suggestion
```

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -503,19 +556,21 @@ void CTurbSASolver::BC_HeatFlux_Wall(CGeometry *geometry, CSolver **solver_conta
-         LinSysRes.SubtractBlock(iPoint, &Res_Wall);
+         su2double Res_Wall[MAXNVAR] = {0.0};
+         Res_Wall[0] = coeff*RoughWallBC*Area;
+         LinSysRes.SubtractBlock(iPoint, Res_Wall);
```
Revert and use `LinSysRes(iPoint, 0) -= Res_Wall;`

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -503,19 +556,21 @@ void CTurbSASolver::BC_HeatFlux_Wall(CGeometry *geometry, CSolver **solver_conta
          su2double coeff = (nu_total/sigma);
          su2double RoughWallBC = nodes->GetSolution(iPoint,0)/(0.03*Roughness_Height);
 
-         su2double Res_Wall;// = new su2double [nVar];
-         Res_Wall = coeff*RoughWallBC*Area;
-         LinSysRes.SubtractBlock(iPoint, &Res_Wall);
+         su2double Res_Wall[MAXNVAR] = {0.0};
+         Res_Wall[0] = coeff*RoughWallBC*Area;
+         LinSysRes.SubtractBlock(iPoint, Res_Wall);
 
-         su2double Jacobian_i = (laminar_viscosity /density *Area)/(0.03*Roughness_Height*sigma);
-         Jacobian_i += 2.0*RoughWallBC*Area/sigma;
-         if (implicit) Jacobian.AddVal2Diag(iPoint, -Jacobian_i);
+         Jacobian_i[0][0] = (laminar_viscosity /density *Area)/(0.03*Roughness_Height*sigma);
+         Jacobian_i[0][0] += 2.0*RoughWallBC*Area/sigma;
```
revert, just your last change to this part is enough to make it compatible right?

AngPass replied:
I think it can’t be reverted, because the Jacobian here is nVar $\times$ nVar, not scalar.

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -1514,17 +1602,297 @@ void CTurbSASolver::SetDES_LengthScale(CSolver **solver, CGeometry *geometry, CC
+  const su2double dummySource = 1e3;
+  bool stochBackscatterInBox = config->GetStochBackscatterInBox();
+  
+  SU2_OMP_FOR_DYN(omp_chunk_size)
```
This has to be around the for not the if
```suggestion
  SU2_OMP_FOR_STAT(omp_chunk_size)
```

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -1514,17 +1602,297 @@ void CTurbSASolver::SetDES_LengthScale(CSolver **solver, CGeometry *geometry, CC
+  }
+  END_SU2_OMP_FOR
+
+  SU2_OMP_FOR_DYN(omp_chunk_size)
```
```suggestion
  SU2_OMP_FOR_STAT(omp_chunk_size)
```

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -1514,17 +1602,297 @@ void CTurbSASolver::SetDES_LengthScale(CSolver **solver, CGeometry *geometry, CC
+        su2double mean_check_new = 0.0;
+        su2double var_check_notSmoothed = 0.0;
+        su2double mean_check_notSmoothed = 0.0;
+        for (unsigned long iPoint = 0; iPoint < nPointDomain; iPoint++) {
```
There is some OpenMP stuff missing; I can help with that later.

pcarruscag commented:
Post some results you obtained with this implementation please.

AngPass committed:
Merge branch 'develop' into develop

AngPass committed:
Update Common/include/toolboxes/random_toolbox.hpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update Common/include/toolboxes/random_toolbox.hpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/src/solvers/CTurbSASolver.cpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/src/solvers/CTurbSASolver.cpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/include/numerics/turbulent/turb_sources.hpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/src/numerics/flow/convection/centered.cpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/src/solvers/CTurbSASolver.cpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass commented:
> Post some results you obtained with this implementation please.

For validation purposes, I simulated the spatially-developing mixing layer. I compared the baseline DDES results against the DDES case with stochastic backscatter enabled. The comparison includes both $Q$‑criterion visualizations and the streamwise evolution of the $XZ$-velocity covariances at two stations downstream of the splitter‑plate trailing edge.
The improvement obtained with the backscatter model is clearly visible in both diagnostics, with enhanced vortex activity and significantly better recovery of the resolved Reynolds stresses. This gives further confidence that the implementation behaves as expected.
The files are listed below:

[qCriterionDDES.pdf](https://github.com/user-attachments/files/26333465/qCriterionDDES.pdf)
[qCriterionDDES_SBS.pdf](https://github.com/user-attachments/files/26333464/qCriterionDDES_SBS.pdf)
[stresses_x_0.15.pdf](https://github.com/user-attachments/files/26333466/stresses_x_0.15.pdf)
[stresses_x_0.2.pdf](https://github.com/user-attachments/files/26333463/stresses_x_0.2.pdf)

AngPass commented:
> Add a regression test to parallel_regression.py, you'll probably need to do something custom to decompress the file. It would be good to make it a bit smaller if possible too.

I’ve added the new regression test to `parallel_regression.py`, including the custom decompression step required for the grid file.
Unfortunately the grid size for this case cannot be reduced further without compromising the physical behavior of the test, so I kept the original mesh.

AngPass commented:
@pcarruscag I’m about to push the second major revision.
For clarity, all conversations that were resolved without an explicit reply have been addressed directly in the code — every requested change has been implemented accordingly.

AngPass committed:
Merge branch 'develop' into develop

Commit:
Second major revision

Commit:
Minor fixes

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -78,56 +78,64 @@ inline double HashToUniform(uint64_t x) {
     double t = 3.75 / abx;
-    double poly = 0.39894228 + t * (0.01328592 + t * (0.00225319 + t * (-0.00157565 + t * (0.00916281 + t * (-0.02057706 + t * (0.02635537 + t * (-0.01647633 + t * 0.00392377)))))));
-    return abx - 0.5 * log(abx) + log(poly);
+    double poly =
```
The intermediate variables also need to be su2double

AngPass replied:
I used a template, following the approach adopted, for example, in `geometry_toolbox.hpp`.

pcarruscag commented on UnitTests/Common/toolboxes/random_toolbox_tests.cpp:
```diff
@@ -0,0 +1,61 @@
+  constexpr double inv53 = 1.0 / 9007199254740992.0;  // 1/2^53
+  uint64_t uInt = x >> 11;                            // top 53 bits
+  return (uInt + 1) * inv53;                          // map to (0,1]
+}
```
Include the header here, don't duplicate the function.

AngPass committed:
Merge branch 'develop' into develop

Commit:
Minor fixes

AngPass commented:
@pcarruscag I implemented the changes you suggested.
I also noticed that CodeFactor is reporting issues which are unrelated to my modifications.

pcarruscag commented on SU2_CFD/include/numerics/turbulent/turb_convection.hpp:
```diff
@@ -59,6 +63,12 @@ class CUpwSca_TurbSA final : public CUpwScalar<FlowIndices> {
+    if (config->GetSBSParam().StochasticBackscatter && config->GetSBSParam().SBS_Ctau > 0.0) {
+      for (unsigned short iVar = 1; iVar < nVar; iVar++) {
+        Flux[iVar] = (a0 + a1) * 0.5 * (ScalarVar_i[iVar] + ScalarVar_j[iVar]);
+        Jacobian_i[iVar][iVar] = 0.5 * (a0+a1);
```
What about Jacobian_j?

AngPass replied:
You’re right. I had overlooked `Jacobian_j`. My apologies for that.

pcarruscag commented on SU2_CFD/include/variables/CTurbSAVariable.hpp:
```diff
@@ -28,6 +28,7 @@
 #pragma once
 
 #include "CTurbVariable.hpp"
+#include <random>
```
```suggestion
```

pcarruscag commented on SU2_CFD/include/numerics/CNumerics.hpp:
```diff
@@ -32,6 +32,7 @@
 #include <iostream>
 #include <limits>
 #include <cstdlib>
+#include <random>
```
```suggestion
```

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -1547,6 +1576,18 @@ void CFlowOutput::SetVolumeOutputFieldsScalarMisc(const CConfig* config) {
+      if (config->GetSBSParam().SBS_Ctau > 0.0) {
+        AddVolumeOutput("STOCHVAR_X", "StochVar_x", "SOLUTION", "x-component of the stochastic vector potential");
+        AddVolumeOutput("STOCHVAR_Y", "StochVar_y", "SOLUTION", "y-component of the stochastic vector potential");
+        if (nDim==3) AddVolumeOutput("STOCHVAR_Z", "StochVar_z", "SOLUTION", "z-component of the stochastic vector potential");
```
You need to order these right after the SA solution otherwise the solver will not restart properly.
Have you checked if the function that reads the restart file needs any tweaks?

AngPass replied:
The ouputs are now ordered immediately after the SA solution to ensure proper restart behavior. I had already updated the restart-reading function to make it compatible with the backscatter formulation. Please, see changes to `CTurbSolver::LoadRestart`.

    unsigned short nVarInRestart = Restart_Vars[1] - skipVars;
    if (nVarInRestart > nVar) nVarInRestart = nVar;

    /*--- Load data from the restart into correct containers. ---*/

    unsigned long counter = 0;
    for (auto iPoint_Global = 0ul; iPoint_Global < geometry[MESH_0]->GetGlobal_nPointDomain(); iPoint_Global++) {
      /*--- Retrieve local index. If this node from the restart file lives
       on the current processor, we will load and instantiate the vars. ---*/

      const auto iPoint_Local = geometry[MESH_0]->GetGlobal_to_Local_Point(iPoint_Global);

      if (iPoint_Local > -1) {
        /*--- We need to store this point's data, so jump to the correct
         offset in the buffer of data from the restart file and load it. ---*/

        const auto index = counter * Restart_Vars[1] + skipVars;
        for (auto iVar = 0u; iVar < nVarInRestart; iVar++) nodes->SetSolution(iPoint_Local, iVar, Restart_Data[index + iVar]);
        for (auto iVar = nVarInRestart; iVar < nVar; iVar++) nodes->SetSolution(iPoint_Local, iVar, 0.0);

This should make it possible to restart from a turbulent solution obtained with backscatter disabled.

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -4051,6 +4154,19 @@ void CFlowOutput::SetTimeAveragedFields() {
+
+  AddVolumeOutput("STOCHASTIC_REYNOLDS_STRESS_12", "StochasticReynoldsStress_12", "TIME_AVERAGE", "Stochastic Reynolds stress xy-component");
+  AddVolumeOutput("STOCHASTIC_REYNOLDS_STRESS_13", "StochasticReynoldsStress_13", "TIME_AVERAGE", "Stochastic Reynolds stress xz-component");
+  AddVolumeOutput("STOCHASTIC_REYNOLDS_STRESS_23", "StochasticReynoldsStress_23", "TIME_AVERAGE", "Stochastic Reynolds stress yz-component");
```
Are the "ii" components 0?

AngPass replied:
Yes, they are. By construction, the stochastic stress tensor is antisymmetric.

pcarruscag commented on SU2_CFD/src/output/CFlowOutput.cpp:
```diff
@@ -4051,6 +4154,19 @@ void CFlowOutput::SetTimeAveragedFields() {
+    AddVolumeOutput("MODELED_REYNOLDS_STRESS_33", "ModeledReynoldsStress_33", "TIME_AVERAGE", "Modeled Reynolds stress zz-component");
+    AddVolumeOutput("MODELED_REYNOLDS_STRESS_13", "ModeledReynoldsStress_13", "TIME_AVERAGE", "Modeled Reynolds stress xz-component");
+    AddVolumeOutput("MODELED_REYNOLDS_STRESS_23", "ModeledReynoldsStress_23", "TIME_AVERAGE", "Modeled Reynolds stress yz-component");
+  }
```
you should guard these with the same conditions used to output then. Otherwise we will write some fields that will all be 0.

AngPass replied:
This has now been addressed. All `nDim` checks associated with the backscatter model have been removed, as backscatter is only meaningful for 3D flows (see the check implemented in `CConfig.cpp`). In addition, the modeled Reynolds stresses can be written regardless of whether backscatter is enabled, as long as the simulation is unsteady; therefore, only the `nDim` check is applied.
Finally, I moved the actual computation of both the modeled and the stochastic stresses into `CFlowOutput::LoadTimeAveragedData`. This required modifications of the member inputs.

pcarruscag commented on SU2_CFD/src/solvers/CSolver.cpp:
```diff
@@ -363,8 +363,8 @@ void CSolver::InitiatePeriodicComms(CGeometry *geometry,
-  auto *Sol_Min   = new su2double[nPrimVarGrad];
-  auto *Sol_Max   = new su2double[nPrimVarGrad];
+  auto *Sol_Min   = new su2double[nVar];
+  auto *Sol_Max   = new su2double[nVar];
```
Why was this changed?

AngPass replied:
This change is related to an issue I encountered which I was unable to solve. Please, refer to the comment below.

> @pcarruscag I am currently encountering a bug that I have not been able to resolve on my own. The issue seems to originate from the handling of MPI communications of periodic limiters. A segmentation fault arises when the backscatter model is active, periodic boundary conditions are present and the flag `MUSCL_TURB` is set to `YES`. I have not identified the root cause yet. Any suggestions or guidance would be greatly appreciated.
> 
> **UPDATE:** it seems that allocating `Sol_Min` and `Sol_Max` in `CSolver::InitiatePeriodicComms` as `auto *Sol_Min = new su2double[nVar]; auto *Sol_Max = new su2double[nVar]` instead of `auto *Sol_Min = new su2double[nPrimVarGrad]; auto *Sol_Max = new su2double[nPrimVarGrad]` solves the issue.

pcarruscag replied:
Hmmm ok makes sense, maybe the turbulence solvers don't define nPrimVarGrad.

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -1514,17 +1594,309 @@ void CTurbSASolver::SetDES_LengthScale(CSolver **solver, CGeometry *geometry, CC
+  auto sbsBoxBounds = config->GetSBSParam().StochBackscatterBoxBounds;
+
+  SU2_OMP_FOR_STAT(omp_chunk_size)
+  for (unsigned long iPoint = 0; iPoint < nPointDomain; iPoint++) {
```
```suggestion
  for (unsigned long iPoint = 0; iPoint < nPoint; iPoint++) {
```

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -1514,17 +1594,309 @@ void CTurbSASolver::SetDES_LengthScale(CSolver **solver, CGeometry *geometry, CC
+    bool outOfBoxY = (coord[1]<sbsBoxBounds[2] || coord[1]>sbsBoxBounds[3]);
+    bool outOfBoxZ = (coord[2]<sbsBoxBounds[4] || coord[2]>sbsBoxBounds[5]);
+    bool outOfBox  = (outOfBoxX || outOfBoxY || outOfBoxZ);
+    if (outOfBox) nodes->SetSbsInBox(iPoint, 0.0);
```
```suggestion
    nodes->SetSbsInBox(iPoint, outOfBox ? 0 : 1);
```

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -1514,17 +1594,309 @@ void CTurbSASolver::SetDES_LengthScale(CSolver **solver, CGeometry *geometry, CC
+          auto* normal = geometry->edges->GetNormal(iEdge);
+          su2double area = GeometryToolbox::Norm(nDim, normal);
+          su2double dx_ij = GeometryToolbox::Distance(nDim, coord_i, coord_j);
+          su2double source_j = nodes->GetLangevinSourceTerms(jPoint, iDim);
```
Shouldn't the smoothing be done on `GetLangevinSourceTermsOld`?
Like this its probably going to be sensitive to MPI partitioning, on the way the points are ordered, etc.

pcarruscag replied:
I suppose it's ok if the iterations converge. I'll keep it in mind for OpenMP

AngPass replied:
`GetLangevinSourceTermsOld` remains fixed throughout the iterations, as it represents the initial uncorrelated random field. In contrast, `GetLangevinSourceTerms` is overwritten at each smoothing iteration.
In the test cases I performed so far, no convergence issues were observed in the smoothing procedure. However, the simulations were performed without OpenMP enabled.

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -1514,17 +1594,309 @@ void CTurbSASolver::SetDES_LengthScale(CSolver **solver, CGeometry *geometry, CC
+            cout << endl;
+          }
+        }
+        break;
```
```suggestion
        /*--- Converged or maximum number of iterations reached. ---*/
        break;
```

pcarruscag commented on SU2_CFD/src/variables/CTurbSAVariable.cpp:
```diff
@@ -47,7 +57,13 @@ CTurbSAVariable::CTurbSAVariable(su2double val_nu_tilde, su2double val_muT, unsi
   Vortex_Tilting.resize(nPoint);
+  stochSource.resize(nPoint, nDim) = su2double(0.0);
+  stochSourceOld.resize(nPoint, nDim) = su2double(0.0);
+  besselIntegral.resize(nPoint) = su2double(0.0);
```
Avoid allocating when it's not needed by the solver please.

AngPass replied:
The allocated variables are actually used by the solver, and at the moment I do not see a viable way to avoid allocating them. I apologize if this is sub‑optimal. Do you have any suggestions?

pcarruscag replied:
Maybe with some if statements to not set them in the numerics classes? Then the solvers will not use them anymore.

AngPass replied:
Oh, yes. Sorry, I had misunderstood your comment. The modifications have been added. Thank you.

pcarruscag reviewed (commented):
Thank you for posting the results, looks very interesting.

AngPass committed:
Merge branch 'develop' into develop

Commit:
Third major revision

AngPass committed:
Update SU2_CFD/include/variables/CTurbSAVariable.hpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/include/numerics/CNumerics.hpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/src/solvers/CTurbSASolver.cpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/src/solvers/CTurbSASolver.cpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/src/solvers/CTurbSASolver.cpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

Commit:
Add residuals of stoch. variables to vol. outputs

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -0,0 +1,164 @@
+/*!
+ * \file random_toolbox.hpp
+ * \brief Collection of utility functions for random number generation.
+ * \version 8.3.0 "Harrier"
```
```suggestion
 * \version 8.4.0 "Harrier"
```

pcarruscag reviewed (commented):
Looks like the compilation is failing in GitHub Actions with some unused variable warnings.
I've invited you to the SU2 organization so that I don't need to approve the workflows, you can just iterate until all the tests run.

AngPass committed:
Update Common/include/toolboxes/random_toolbox.hpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Merge branch 'develop' into develop

Commit:
Fix compilation errors

Commit:
Fix compilation errors - bis

Commit:
Fix compilation errors - tris

Commit:
Fix compilation errors - tetra

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -0,0 +1,180 @@
+
+  if (abx < 3.75) {
+    auto t = abx / 3.75;
+    auto p =
```
We can't use auto on the results of expressions of active types.
```suggestion
inline T GetBesselZero(const T& x) {
  T abx = fabs(x);

  if (abx < 3.75) {
    T t = abx / 3.75;
    T p =
```

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -0,0 +1,180 @@
+                                      t * (-0.01647633 +
+                                           t * 0.00392377)))))));
+
+    auto arg = sqrt(abx) * poly;
```
```suggestion
    T t = 3.75 / abx;

    T poly =
        0.39894228 +
        t * (0.01328592 +
             t * (0.00225319 +
                  t * (-0.00157565 +
                       t * (0.00916281 +
                            t * (-0.02057706 +
                                 t * (0.02635537 +
                                      t * (-0.01647633 +
                                           t * 0.00392377)))))));

    T arg = sqrt(abx) * poly;
```

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -0,0 +1,180 @@
+  const auto A = T(1.0) + T(2.0) * (beta_x + beta_y + beta_z);
+  const auto Bx = T(2.0) * beta_x;
+  const auto By = T(2.0) * beta_y;
+  const auto Bz = T(2.0) * beta_z;
```
```suggestion
template <class T>
inline T GetBesselIntegral(const T& beta_x, const T& beta_y, const T& beta_z) {
  const T A = 1 + 2 * (beta_x + beta_y + beta_z);
  const T Bx = 2 * beta_x;
  const T By = 2 * beta_y;
  const T Bz = 2 * beta_z;
```

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -0,0 +1,180 @@
+  const auto Bz = T(2.0) * beta_z;
+
+  const int N = 4000;
+  const auto t_max = T(20.0);
```
```suggestion
  const T t_max = 20.0;
```

pcarruscag commented on Common/include/toolboxes/random_toolbox.hpp:
```diff
@@ -0,0 +1,180 @@
+
+    auto integrand = exp(lin);
+
+    auto weight = (i == N) ? T(0.5) : T(1.0);
```
```suggestion
  const T dt = t_max / N;

  T sum = T(0.0);

  for (int i = 1; i <= N; i++) {
    T t = i * dt;

    T lin = log(t) - A * t + GetBesselZero(T(Bx * t)) + GetBesselZero(T(By * t)) + GetBesselZero(T(Bz * t));

    T integrand = exp(lin);

    T weight = (i == N) ? 0.5 : 1.0;
```

pcarruscag committed:
Apply suggestions from code review

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

pcarruscag commented on SU2_CFD/include/variables/CTurbSAVariable.hpp:
```diff
@@ -40,7 +40,12 @@ class CTurbSAVariable final : public CTurbVariable {
 private:
   VectorType DES_LengthScale;
+  VectorType lesMode;
+  VectorType sbsInBox;
```
```suggestion
  su2vector<int8_t> sbsInBox;
```

pcarruscag commented on SU2_CFD/include/variables/CTurbSAVariable.hpp:
```diff
@@ -73,6 +78,63 @@ class CTurbSAVariable final : public CTurbVariable {
+   * \param[in] iPoint - Point index.
+   * \param[in] val_sbsInBox - 1.0 if the Stochastic Backscatter Model must be applied only in a confined box.
+   */
+  inline void SetSbsInBox(unsigned long iPoint, su2double val_sbsInBox) override { sbsInBox(iPoint) = val_sbsInBox; }
```
```suggestion
  inline void SetSbsInBox(unsigned long iPoint, int8_t val_sbsInBox) override { sbsInBox(iPoint) = val_sbsInBox; }
```

pcarruscag commented on SU2_CFD/include/variables/CTurbSAVariable.hpp:
```diff
@@ -73,6 +78,63 @@ class CTurbSAVariable final : public CTurbVariable {
+   * \param[in] iPoint - Point index.
+   * \return 1.0 if the Stochastic Backscatter Model must be applied only in a confined box.
+   */
+  inline su2double GetSbsInBox(unsigned long iPoint) const override { return sbsInBox(iPoint); }
```
```suggestion
  inline int8_t GetSbsInBox(unsigned long iPoint) const override { return sbsInBox(iPoint); }
```

pcarruscag commented on SU2_CFD/include/variables/CTurbVariable.hpp:
```diff
@@ -111,4 +111,32 @@ class CTurbVariable : public CScalarVariable {
+   * \param[in] iPoint - Point index.
+   * \param[in] val_sbsInBox - 1.0 if the Stochastic Backscatter Model must be applied only in a confined box.
+   */
+  inline virtual void SetSbsInBox(unsigned long iPoint, su2double val_sbsInBox) {}
```
```suggestion
  inline virtual int8_t GetSbsInBox(unsigned long iPoint) const { return 0; }

  /*!
   * \brief A virtual member.
   * \param[in] iPoint - Point index.
   * \param[in] val_sbsInBox - 1.0 if the Stochastic Backscatter Model must be applied only in a confined box.
   */
  inline virtual void SetSbsInBox(unsigned long iPoint, int8_t val_sbsInBox) {}
```

pcarruscag commented on SU2_CFD/src/variables/CTurbSAVariable.cpp:
```diff
@@ -48,6 +58,17 @@ CTurbSAVariable::CTurbSAVariable(su2double val_nu_tilde, su2double val_muT, unsi
+  if (config->GetKind_HybridRANSLES() != NO_HYBRIDRANSLES) {
+    lesMode.resize(nPoint) = su2double(0.0);
+    if (backscatter) {
+      sbsInBox.resize(nPoint) = su2double(1.0);
```
```suggestion
      sbsInBox.resize(nPoint) = 1;
```

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -378,6 +416,15 @@ void CTurbSASolver::Source_Residual(CGeometry *geometry, CSolver **solver_contai
+        for (unsigned short iDim = 0; iDim < nDim; iDim++)
+          numerics->SetStochSource(nodes->GetLangevinSourceTerms(iPoint, iDim), iDim);
+        numerics->SetLES_Mode(nodes->GetLES_Mode(iPoint), 0.0);
+        numerics->SetSbsInBoxSensor(std::nearbyint(nodes->GetSbsInBox(iPoint)), 0);
```
```suggestion
        numerics->SetSbsInBoxSensor(nodes->GetSbsInBox(iPoint), 0);
```

pcarruscag reviewed (commented):
For the compilation errors

AngPass committed:
Update SU2_CFD/include/variables/CTurbSAVariable.hpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/include/variables/CTurbSAVariable.hpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/include/variables/CTurbSAVariable.hpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/include/variables/CTurbVariable.hpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/src/variables/CTurbSAVariable.cpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

AngPass committed:
Update SU2_CFD/src/solvers/CTurbSASolver.cpp

Co-authored-by: Pedro Gomes <38071223+pcarruscag@users.noreply.github.com>

Commit:
Fix code formatting

Commit:
Fix compilation errors - V

Commit:
Fix errors in regression tests (hopefully)

Commit:
Restore nPrimVarGrad in InitiatePeriodicComms

Commit:
Restore DDES test case (committed by mistake)

pcarruscag commented:
We have to simplify your test, fewer timesteps for example.
As it it is it more than doubles the time to run parallel_regression.py
```
test duration: 16.83 min
```

AngPass committed:
Merge branch 'develop' into develop

Commit:
Simplify backscatter test

AngPass commented on SU2_CFD/src/solvers/CSolver.cpp:
```diff
@@ -363,8 +363,8 @@ void CSolver::InitiatePeriodicComms(CGeometry *geometry,
   auto *Und_Lapl  = new su2double[nVar];
-  auto *Sol_Min   = new su2double[nVar];
-  auto *Sol_Max   = new su2double[nVar];
+  auto *Sol_Min   = new su2double[nPrimVarGrad];
+  auto *Sol_Max   = new su2double[nPrimVarGrad];
```
It appears that using `nVar` leads to segmentation fault errors in test cases employing the $k-\omega$ turbulence model with MUSCL reconstruction and periodic boundary conditions. Reverting to the original version of the file (which uses `nPrimVarGrad`) resolves this issue, but then a segmentation fault occurs when running cases with the Spalart–Allmaras model coupled with backscatter, MUSCL reconstruction, and periodic boundaries. I have not been able to identify the root cause of these errors.

AngPass committed:
Merge branch 'develop' into develop

pcarruscag committed:
Merge branch 'develop' into develop

AngPass committed:
Merge branch 'develop' into develop

Commit:
Fix compilation issues

Commit:
Fix parallel regression

pcarruscag commented:
We can merge this and I'll address the OpenMP comment. Can you update all the failling regressions please?
We started requiring a linear history for the develop branch. To that end you can either rebase this PR on develop, which will preserve all commits, or we can squash merge, which will make a single commit for the whole thing.
Let me know how you want to proceed.

Commit:
Fix regressions

AngPass committed:
Merge branch 'develop' into develop

Commit:
Fix regressions #2

AngPass commented:
@pcarruscag I updated the failing regressions. 
Anyway, squash merge is fine for me. Thanks!
